_posts/2015-08-15-Cocoa-Architecture:-Router-Pattern.markdown

```diff
@@ -0,0 +1,164 @@
+---
+layout: post
+title: "Cocoa Architecture: ARRouter"
+date: 2015-08-15
+comments: false
+categories: [ios, mobile, architecture, energy, eigen, eidolon, hal]
+author: Orta Therox
+github-url: https://www.github.com/orta
+twitter-url: http://twitter.com/orta
+blog-url: http://orta.io
+---
+
+I want to talk about a pattern that we've been using for the last few years on the Artsy Mobile team. This pattern pre-dates me joining Artsy by a few weeks, and was introduced into our codebase by Ben Jackson, this was it's first method:
+
+```objc
+  + (NSURL *)newOAuthURLWithUsername:(NSString *)username password:(NSString *)password {
+      NSDictionary *params = [[NSDictionary alloc] initWithObjectsAndKeys:
+                              username, @"email",
+                              password, @"password",
+                              ARAuthClientID, @"client_id",
+                              ARAuthSecret, @"client_secret",
+                              @"credentials", @"grant_type",
+                              nil];
+      NSString *url_string = [[NSString alloc] initWithFormat:@"%@%@", AROAuthURL, [params queryString]];
+      NSURL *url = [ARRouter newURLWithPath:url_string];
+      [url_string release];
+      [params release];
+      return url;
+  }
+```
+
+Yep, that's pre-ARC, pre-Dictionary Literals, memory-managed code. We took this pattern and rolled with it for the next 4 years, this article is about where we took it.
```

Comment:
Awesome intro! Maybe it should end with "this article is about where we've taken it"?